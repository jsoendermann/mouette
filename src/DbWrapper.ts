import { MongoClient, Db } from 'mongodb'
import { flatten, uniq } from 'lodash'


export class DbWrapper {
  private connectionPromise: Promise<Db>

  constructor(private mongoUri: string) { }

  private async getDb(): Promise<Db> {
    if (this.connectionPromise) {
      return this.connectionPromise
    }

    this.connectionPromise = MongoClient.connect(this.mongoUri).then((db: Db) => {
      if (!db) {
        delete this.connectionPromise;
        throw new Error('db is falsy')
      }

      db.on('error', () => delete this.connectionPromise)
      db.on('close', () => delete this.connectionPromise)

      return db
    }).catch(error => {
      delete this.connectionPromise
      return Promise.reject(error)
    })

    return this.connectionPromise
  }

  public async close() {
    const db = await this.connectionPromise
    db.close()
  }

  private collectionNames: Promise<string[]> | null = null
  public async getCollectionNames(): Promise<string[]> {
    if (this.collectionNames) {
      return this.collectionNames
    }

    this.collectionNames = this.getDb().then(async db => {
      const collectionDetails = await db.listCollections({}).toArray()
      return collectionDetails
        .map(d => d.name)
        .filter(n => !n.startsWith('system.'))
    })

    return this.collectionNames
  }

  private collectionToKeys: Map<string, Promise<string[]>> = new Map()
  public async getKeysInCollection(collectionName: string): Promise<string[]> {
    const keysInCollection = this.collectionToKeys.get(collectionName)

    if (keysInCollection) {
      return keysInCollection
    }

    const promise = this.getDb().then(async db => {
      const result = await db.collection(collectionName).mapReduce(
        'function () { for (var key in this) { emit(key, null) }}',
        'function () {}',
        { out: { inline: 1 } }
      )

      return result.map((k: any) => k._id) as string[]
    })

    this.collectionToKeys.set(collectionName, promise)
    return promise
  }
}