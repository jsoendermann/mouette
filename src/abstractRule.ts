import { Db } from 'mongodb'


export interface IRuleFailureJson {
  ruleName: string
  prettyRuleName: string
  granularity: string
  location: {
    collectionName?: string
    keyName?: string
    recordId?: string
  }
  isFuzzy: boolean
  severity: string
  failure: string
  fix?: string
  hash: string
}

export interface IRuleFailure {
  toJson(): IRuleFailureJson
  getCollectionName(): string | undefined
  getKeyName(): string | undefined
  getRecordId(): any | undefined
}

export class RuleFailure implements IRuleFailure {
  constructor(private rule: IRule,
              private collectionName?: string,
              private keyName?: string,
              private recordId?: any) { }
    
  public toJson(): IRuleFailureJson {
    return this.rule.failureToJson(this)
  }

  getCollectionName(): string | undefined {
    return this.collectionName
  }

  getKeyName(): string | undefined {
    return this.keyName
  }

  getRecordId(): any | undefined {
    return this.recordId
  }
}

export type RuleSeverity = 'warning' | 'error'
export type RuleGranularity = 'collection_name' | 'key_name' | 'column' | 'row'

export interface IRuleMetadata {
  name: string
  prettyName: string
  description: string
  rationale: string
  severity: RuleSeverity
  granularity: RuleGranularity
  isFuzzy: boolean
}


export interface IRule {
  readonly metadata: IRuleMetadata
  apply(db: Db): Promise<IRuleFailure[]>
  failureToJson(failure: IRuleFailure): IRuleFailureJson
}
