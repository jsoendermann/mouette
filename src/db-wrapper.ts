import { Db, MongoClient } from 'mongodb'


export class DbWrapper {
  private connectionPromise: Promise<Db>

  constructor(private mongoUrl: string) {}

  private async getDb(): Promise<Db> {
    if (this.connectionPromise) {
      return this.connectionPromise
    }

    this.connectionPromise = MongoClient.connect(this.mongoUrl).then((db: Db) => {
      if (!db) {
        delete this.connectionPromise
        throw new Error('db is falsy')
      }

      db.on('error', () => delete this.connectionPromise)
      db.on('close', () => delete this.connectionPromise)

      return db
    }).catch(error => {
      delete this.connectionPromise
      return Promise.reject(error)
    })

    return this.connectionPromise
  }

}