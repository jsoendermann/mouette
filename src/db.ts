import { Db, MongoClient } from 'mongodb'


export type MongoType =
  | 'missing'
  | 'null'
  | 'boolean'
  | 'number'
  | 'string'
  | 'array'
  | 'object'
  | 'ObjectId'
  | 'Date'

export interface IDb {
  /**
   * Returns the names of all collections in the db. Calling this method is very cheap since it caches the result.
   *
   * @returns {Promise<string[]>} A promise that resolves to an array containing the collection names.
   *
   * @memberOf IDb
   */
  getCollectionNames(): Promise<string[]>

  /**
   * Returns the union of the keys of all documents in the collection. Calling this method is very cheap
   * since it caches the result.
   *
   * @param {string} collectionName The collection for which you want to get key names.
   * @returns {Promise<string[]>} A promise that resolves to an array containing the key names.
   *
   * @memberOf IDb
   */
  getKeysInCollection(collectionName: string): Promise<string[]>

  /**
   * Returns the types of the values found in the given column. Calling this method is very cheap since it
   * caches the result.
   *
   * @param {string} collectionName The collection that contains the column.
   * @param {string} keyName The name of the column.
   * @returns {Promise<MongoType[]>} A promise that resolves to an array containing all the MongoTypes that
   * are found in the column.
   *
   * @memberOf IDb
   */
  getTypesOfKeyInCollection(collectionName: string, keyName: string): Promise<MongoType[]>

  /**
   * Whether any documents satisfying the given query exist in the collection.
   *
   * @param {string} collectionName The name of the collection.
   * @param {*} query A MongoDB query
   * @returns {Promise<boolean>} A promise that resolves to a boolean indicating whether the collection doe
   * contain at least one document that satisfies the query.
   *
   * @memberOf IDb
   */
  doesContainInCollection(collectionName: string, query: any): Promise<boolean>

  /**
   * Closes the connection to the db.
   *
   * @returns {Promise<void>} A promise that gets resolved when the connection has been closed.
   *
   * @memberOf IDb
   */
  close(): Promise<void>
}

export class MongoDbWrapper implements IDb {
  private connectionPromise: Promise<Db>

  // Used for caching
  private collectionNames: Promise<string[]> | null = null
  private collectionToKeys: Map<string, Promise<string[]>> = new Map()
  private collectionToKeyToTypes: Map<string, Promise<Map<string, MongoType[]>>> = new Map()

  constructor(private mongoUrl: string) { }

  private async getDb(): Promise<Db> {
    if (this.connectionPromise) {
      return this.connectionPromise
    }

    this.connectionPromise = MongoClient.connect(this.mongoUrl).then((db: Db) => {
      if (!db) {
        delete this.connectionPromise
        throw new Error('db is falsy')
      }

      db.on('error', () => delete this.connectionPromise)
      db.on('close', () => delete this.connectionPromise)

      return db
    }).catch(error => {
      delete this.connectionPromise
      return Promise.reject(error)
    })

    return this.connectionPromise
  }

  public async getCollectionNames(): Promise<string[]> {
    if (this.collectionNames) {
      return this.collectionNames
    }

    this.collectionNames = this.getDb().then(async db => {
      const collectionDetails = await db.listCollections({}).toArray()
      return collectionDetails
        .map(d => d.name)
        .filter(n => !n.startsWith('system.'))
    })

    return this.collectionNames
  }

  public async getKeysInCollection(collectionName: string): Promise<string[]> {
    const keysInCollection = this.collectionToKeys.get(collectionName)

    if (keysInCollection) {
      return keysInCollection
    }

    const promise = this.getDb().then(async db => {
      const result = await db.collection(collectionName).mapReduce(
        'function () { for (var key in this) { emit(key, null) } }',
        'function () {}',
        { out: { inline: 1 } },
      )

      return result.map((k: any) => k._id) as string[]
    })

    this.collectionToKeys.set(collectionName, promise)
    return promise
  }

  private async getTypesOfKeysInCollection(
    collectionName: string,
  ): Promise<Map<string, MongoType[]>> {
    const keyToType = this.collectionToKeyToTypes.get(collectionName)

    if (keyToType) {
      return keyToType
    }

    const promise = this.getDb().then(async db => {
      const keys = await this.getKeysInCollection(collectionName)
      const result = await db.collection(collectionName).mapReduce(
        `
      function() {
        var keys = ["${keys.join('","')}"];
        for (var i = 0; i < keys.length; i++) {
          var key = keys[i];
          var value = this[key];
          var result;
          if (value instanceof ObjectId) {
            result = {'ObjectId': true};
          } else if (value instanceof Date) {
            result = {'Date': true};
          } else if (value === null) {
            result = {'null': true};
          } else if (value === undefined) {
            result = {'missing': true};
          } else {
            var type = typeof value;
            result = {};
            result[type] = true;
          }
          emit(key, result)
        }
      }
      `
        ,
        `
      function (key, values) {
        var result = {};
        for (var i = 0; i < values.length; i++) {
          var value = values[i];
          for (var type in value) {
            if (value.hasOwnProperty(type) && value[type]) {
              print(type)
              result[type] = true;
            }
          }
        }
        return result;
      }
      `,
        { out: { inline: 1 } },
      )

      const map = new Map<string, MongoType[]>()
      result.forEach(({ _id, value }: { _id: string, value: any }) => {
        const types: MongoType[] = []
        Object.keys(value).forEach(type => types.push(type as MongoType))
        map.set(_id, types)
      })

      return map
    })

    this.collectionToKeyToTypes.set(collectionName, promise)

    return promise
  }

  public async getTypesOfKeyInCollection(
    collectionName: string,
    keyName: string,
  ): Promise<MongoType[]> {
    const types = this
      .getTypesOfKeysInCollection(collectionName)
      .then(keysToType => keysToType.get(keyName)!)
    return types
  }

  public async doesContainInCollection(collectionName: string, query: any): Promise<boolean> {
    const db = await this.getDb()
    const record = await db.collection(collectionName).findOne(query)
    return !!record
  }

  public async close() {
    const db = await this.connectionPromise
    db.close()
  }
}


export class TestDbWrapper implements IDb {
  public async getCollectionNames(): Promise<string[]> {
    throw new Error('getCollectionNames should be mocked')
  }

  public async getKeysInCollection(collectionName: string): Promise<string[]> {
    throw new Error('getKeysInCollection should be mocked')
  }

  public async getTypesOfKeyInCollection(
    collectionName: string,
    keyName: string,
  ): Promise<MongoType[]> {
    throw new Error('getTypesOfKeyInCollection should be mocked')
  }

  public async doesContainInCollection(collectionName: string, query: any): Promise<boolean> {
    throw new Error('doesContainInCollection should be mocked')
  }

  public async close(): Promise<void> {
    throw new Error('close should be mocked')
  }
}
