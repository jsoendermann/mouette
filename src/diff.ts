import { differenceBy } from 'lodash'

import { IRuleFailureJson } from './rule'


export const diff = (
  lintResultsOld: IRuleFailureJson[],
  lintResultsNew: IRuleFailureJson[],
): IRuleFailureJson[] => {
  return differenceBy(
    lintResultsNew,
    lintResultsOld,
    (a: IRuleFailureJson) => `${
      a.ruleMetadata.name}.${
      a.location.collectionName}.${
      a.location.keyName}`,
  )
}
