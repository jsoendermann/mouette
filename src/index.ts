export * from './run'
export * from './diff'
