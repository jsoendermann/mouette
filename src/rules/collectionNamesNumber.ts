import { DbWrapper } from '../DbWrapper'
import { AbstractRule, IRuleFailure, IRuleFailureJson, RuleFailure,  RuleGranularity, RuleSeverity } from '../rule'
import { isPlural, pluralize } from '../vendor/lingo'


export class Rule extends AbstractRule {
  public static metadata = {
    name: 'collection-names-number',
    prettyName: 'Collection names number',
    description: 'Make sure all collections in the database have names that are pluralized.',
    rationale: 'Reads more fluently.',
    severity: 'warning' as RuleSeverity,
    granularity: 'collection_name' as RuleGranularity,
    isFuzzy: false,
    options: {

    },
  }

  public async apply(db: DbWrapper): Promise<IRuleFailure[]> {
    const collectionNames = await db.getCollectionNames()
    const nonPluralizedNames = collectionNames.filter(name => !isPlural(name))
    return nonPluralizedNames.map(name => new RuleFailure(this, name))
  }

  public failureToJson(failure: IRuleFailure): IRuleFailureJson {
    const collectionName = failure.getCollectionName() as string

    return {
      ruleMetadata: Rule.metadata,
      location: {
        collectionName,
      },
      failure: `Collection name **${collectionName}** is not pluralized.`,
      suggestion: `Change *${collectionName}* to *${pluralize(collectionName)}*.`,
    }
  }
}
