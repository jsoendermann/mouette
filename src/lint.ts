import { readFileSync } from 'fs'
import * as humps from 'humps'
import { merge, flatten } from 'lodash'
import { dirname, resolve } from 'path'
import { parse } from 'toml'

import { MongoDbWrapper } from './db'
import { IRuleFailureJson, AbstractRule } from './rule'


export const lint = async (mongoUrl: string, userConfig: any = {}): Promise<IRuleFailureJson[]> => {
  const db = new MongoDbWrapper(mongoUrl)

  const moduleFolder = dirname(module.filename)
  const defaultConfigFileName = resolve(moduleFolder, '../defaultConfig.toml')
  const defaultConfigFileContents = readFileSync(defaultConfigFileName, 'utf8')
  const defaultConfig = parse(defaultConfigFileContents)

  const config = merge(defaultConfig, userConfig)

  const rules: AbstractRule[] = []

  for (const ruleName of Object.keys(config)) {
    const { enabled, ...ruleOptions } = config[ruleName]
    if (enabled) {
      const ruleFileName = resolve(`${dirname(module.filename)}/rules/${humps.camelize(ruleName)}`)
      const Rule = require(ruleFileName).Rule
      const rule = new Rule(ruleOptions)
      rules.push(rule)
    }
  }

  const failures = flatten(await Promise.all(rules.map(r => r.getFailures(db))))
  const failuresJson = failures.map(f => f.toJson())
  db.close()
  return failuresJson
}
