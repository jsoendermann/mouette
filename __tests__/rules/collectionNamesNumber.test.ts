import { Rule } from '../../src/rules/collectionNamesNumber'
jest.mock('../../src/db')
import { TestDbWrapper } from '../../src/db'

describe('collection-names-number', () => {
  it('should have correct metadata', async () => {
    const rule = new Rule({
      number: 'plural',
    })

    const m = rule.getMetadata()

    expect(m).toMatchSnapshot()
  })

  it("should work when number is 'singular'", async () => {
    const db = new TestDbWrapper();

    (db.getCollectionNames as jest.Mock<any>).mockReturnValueOnce(['cow', 'cats'])

    const rule = new Rule({
      number: 'singular',
    })

    const result = await rule.apply(db)

    expect(result).toMatchSnapshot()
  })

  it("should work when number is 'plural'", async () => {
    const db = new TestDbWrapper();

    (db.getCollectionNames as jest.Mock<any>).mockReturnValueOnce(['cow', 'cats'])

    const rule = new Rule({
      number: 'plural',
    })

    const result = await rule.apply(db)

    expect(result).toMatchSnapshot()
  })
})