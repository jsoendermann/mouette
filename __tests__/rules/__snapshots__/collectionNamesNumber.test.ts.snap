// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collection-names-number should have correct metadata 1`] = `
Object {
  "description": "Make sure all collections in the database have names that are pluralized.",
  "granularity": 0,
  "isFuzzy": false,
  "name": "collection-names-number",
  "optionsDescription": "
      [number]
        description = \\"The grammatical number that collection names should be checked for.\\"
        type = \\"enum\\"
        possibleValues = \\"'singular' | 'plural'\\"
    ",
  "optionsSchema": Object {
    "number":  {
      "_description": null,
      "_examples": Array [],
      "_flags": Object {
        "presence": "required",
      },
      "_inner": Object {},
      "_invalids": Set {
        "_set": Array [],
      },
      "_meta": Array [],
      "_notes": Array [],
      "_refs": Array [],
      "_settings": null,
      "_tags": Array [],
      "_tests": Array [],
      "_type": "any",
      "_unit": null,
      "_valids": Set {
        "_set": Array [
          "singular",
          "plural",
        ],
      },
      "isJoi": true,
    },
  },
  "prettyName": "Collection names number",
  "rationale": "Reads more fluently.",
  "severity": 0,
}
`;

exports[`collection-names-number should work when number is 'singular' 1`] = `
Array [
  RuleFailure {
    "collectionName": "cats",
    "keyName": undefined,
    "rule": Rule {
      "options": Object {
        "number": "singular",
      },
    },
  },
]
`;
